Add support for pubkey auth for users whose home dir is on a NFS share with root squash.

--- svr-auth.c.orig
+++ svr-auth.c
@@ -205,7 +205,26 @@
 			strncmp(methodname, AUTH_METHOD_PUBKEY,
 				AUTH_METHOD_PUBKEY_LEN) == 0) {
 		if (valid_user) {
+			uid_t uid;
+			gid_t gid;
+
+			/* Perform pubkey authentication as the user.
+			 * Pubkey authentication reads user's keys from user's home dir.
+			 * Doing this as the user allows authentication even if user's
+			 * home dir is on an NFS share with root squash attribute. */
+			uid = getuid();
+			gid = getgid();
+			if ((setegid(ses.authstate.pw_gid)) < 0 ||
+				(seteuid(ses.authstate.pw_uid)) < 0) {
+				dropbear_exit("Failed to set euid");
+			}
+
 			svr_auth_pubkey();
+
+			if ((seteuid(uid)) < 0 ||
+				(setegid(gid)) < 0) {
+				dropbear_exit("Failed to revert euid");
+			}
 		} else {
 			/* pubkey has no failure delay */
 			send_msg_userauth_failure(0, 0);
